spec/requests/admin_payroll_run_downloads_spec.rb

```diff
@@ -12,5 +12,67 @@
       expect(response).to have_http_status(:ok)
       expect(response.body).to include("#{Date.today.strftime("%B")} payroll file")
     end
+
+    it "redirects if all payroll_runs have been downloaded" do
+      sign_in_to_admin_with_role(AdminSession::SERVICE_OPERATOR_DFE_SIGN_IN_ROLE_CODE)
+
+      payroll_run = create(:payroll_run, downloaded_at: Time.zone.now)
+
+      expect(get(new_admin_payroll_run_download_path(payroll_run))).to redirect_to admin_payroll_run_download_path(payroll_run)
+    end
+  end
+
+  describe "downloads#show" do
+  it "redirects to the new action when payroll run has not been marked as downloaded" do
+    payroll_run = create(:payroll_run)
+
+    expect(get(admin_payroll_run_download_path(payroll_run))).to redirect_to new_admin_payroll_run_download_path(payroll_run)
+  end
+
+  it "shows a link to the download the payroll run file when payroll run has been marked as downloaded" do
+    payroll_run = create(:payroll_run, downloaded_at: Time.zone.now, downloaded_by: admin_session_id)
+
+    get admin_payroll_run_download_path(payroll_run)
+
+    expect(response.body).to include "Download the payroll run file for #{Date.today.strftime("%B")}."
+    expect(response.body).to include admin_payroll_run_download_path(payroll_run, format: :csv)
+  end
+
+  it "shows when and which user download the payroll run file when payroll run has been marked as downloaded" do
+    payroll_run = create(:payroll_run, downloaded_at: Time.zone.now, downloaded_by: "admin_user_id")
+
+    travel_to 31.seconds.from_now do
```

Comment:
What was your opinion on this one?

Reply:
I don't think you could because the first time is the time at whicch the fake payroll run was triggered for download and we need to travel to 31 seconds later to check that the view has changed and the right content is shown.

If we set the downloaded_at to the same the view would be different.

Reply:
I'm not sure I follow. If we don't "travel forward" in time, but we pretend that the payroll run was created 31 seconds ago, doesn't that allow you to test the same thing?

Reply:
Sorry, got you! Yep, you are of course correct! 👍

Reply:
One difference between the two approaches that's worth being aware of: using time travel locks the time to the time you travel to which means it is deterministic in the specs. The clock will still be running if the timestamp is changed, which may or may not be a problem depending on the specifics of the spec.

Reply:
Ah, good point.